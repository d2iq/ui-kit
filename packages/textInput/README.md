# Forms

## TextInput
`TextInput` is a component used to enter text for forms.

### TextInput Appearance & States

**Appearance**
 
 - **Standard**: The default `appearance` prop value is `TextInputAppearance.Standard`.
 - **Error**: The `appearance` prop can be set to `TextInputAppearance.Error`  to render the `TextInput` as invalid. This should be used if the current value has been confirmed invalid.
 - **Success**: The `appearance` prop can be set to `TextInputAppearance.Success` to render the `TextInput` as validated. This should be used if the current value has confirmed valid.

**States**
 
 - **Disabled**: setting the `disabled` prop to `true` will put the `TextInput` into a disabled state. The `TextInput` will be rendered with a grey background and will not be able to take input.
 
### TextInput `inputLabel`

The `inputLabel` prop can be given either a `string` or `ReactNode` to render as a label for the `TextInput` component.

The `showInputLabel` prop can be set to `false` to hide the input's label. This prop defaults to `true`. 

### TextInput `tooltipContent`

The `tooltipContent` prop is a non-required property that can be a `ReactNode`.

If the `tooltipContent` prop is set, an icon tooltip with the given text will be shown next to the label and above the input.

## TextInputWithIcon
`TextInputWithIcon` extends the `TextInput` component but adds two additional props, `iconStart` & `iconEnd`. These props can be provided a `ReactNode` to be rendered either at the start or end of the `TextInput` or both.

## TextInputWithBadges
`TextInputWithBadges` extends the `TextInputWithIcon` component. More info coming soon...

## TextInputWithButtons
`TextInputWithButtons` extends the `TextInputWithIcon` component, does not take an `iconEnd` prop.

A `TextInputWithButtons` component is used when there is an action a user can take that is related to the text input. For example, clicking a button to clear the input's value.

Please do not pass more than 2 buttons into a `TextInputWithButtons`. Instead, consider putting additional buttons outside of the text input.

### Buttons
The `TextInputButton` component is provided to keep text input buttons consistent. To save space, the buttons are always icons. These icon buttons can be filled with any color, and they do not get the same fill color as icons inside the text input that are not buttons.

Because these buttons contain no text, please pass an `aria-label` to the button. For example, a button that clears the text input might get `aria-label="Clear input"`.
