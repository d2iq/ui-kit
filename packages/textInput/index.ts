export { default as TextInput } from "./components/TextInput";
export { default as TextInputWithIcon } from "./components/TextInputWithIcon";
export { default as TextInputWithBadges } from "./components/TextInputWithBadges";
export { default as TextInputWithButtons } from "./components/TextInputWithButtons";
