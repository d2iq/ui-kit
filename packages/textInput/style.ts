import { css } from "@emotion/css";
import { spaceXs } from "../design-tokens/build/js/designTokens";
import { textInputHeight } from "../shared/styles/formStyles";

const badgeGapSize = spaceXs;

export const badgeInputContainer = css`
  height: auto;
  min-height: ${textInputHeight}px;
  padding-bottom: ${parseInt(badgeGapSize, 10) / 2}px;
  padding-top: ${parseInt(badgeGapSize, 10) / 2}px;
`;

export const badgeText = css`
  white-space: nowrap;
`;

export const badgeInputContents = css`
  display: inline-flex;
  flex-direction: column;
  justify-content: center;
  padding-bottom: ${parseInt(badgeGapSize, 10) / 2}px;
  padding-top: ${parseInt(badgeGapSize, 10) / 2}px;
  padding-right: ${badgeGapSize};

  &,
  * {
    max-width: 100%;
  }
`;

export const badgeInput = css`
  flex-grow: 1;
  min-width: 50px;
`;
