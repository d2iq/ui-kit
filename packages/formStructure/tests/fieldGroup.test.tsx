import * as React from "react";

import { FieldGroup } from "..";
import { TextInput } from "../../textInput";
import { render } from "@testing-library/react";

describe("FieldGroup", () => {
  it("renders", () => {
    const { asFragment } = render(
      <FieldGroup
        direction={{
          default: "column",
          small: "row"
        }}
      >
        <TextInput id="1" />
        <TextInput id="2" />
        <TextInput id="3" />
      </FieldGroup>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
