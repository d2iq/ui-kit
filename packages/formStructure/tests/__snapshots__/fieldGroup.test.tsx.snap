// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldGroup renders 1`] = `
<DocumentFragment>
  <div
    class="css-x7zrii"
    data-cy="flex"
  >
    <div
      class="css-11labrq"
      data-cy="flexItem"
    >
      <div
        data-cy="textInput textInput.standard"
      >
        <label
          class="css-1lrhbps"
          data-cy="textInput-label"
          for="1"
        />
        <div>
          <div
            class="css-865tln"
          >
            <input
              aria-describedby=""
              aria-invalid="false"
              class="css-1w8it3t"
              data-cy="textInput-input"
              id="1"
              type="text"
              value=""
            />
          </div>
          <div
            data-cy="textInput-hintContent"
          />
        </div>
      </div>
    </div>
    <div
      class="css-11labrq"
      data-cy="flexItem"
    >
      <div
        data-cy="textInput textInput.standard"
      >
        <label
          class="css-1lrhbps"
          data-cy="textInput-label"
          for="2"
        />
        <div>
          <div
            class="css-865tln"
          >
            <input
              aria-describedby=""
              aria-invalid="false"
              class="css-1w8it3t"
              data-cy="textInput-input"
              id="2"
              type="text"
              value=""
            />
          </div>
          <div
            data-cy="textInput-hintContent"
          />
        </div>
      </div>
    </div>
    <div
      class="css-11labrq"
      data-cy="flexItem"
    >
      <div
        data-cy="textInput textInput.standard"
      >
        <label
          class="css-1lrhbps"
          data-cy="textInput-label"
          for="3"
        />
        <div>
          <div
            class="css-865tln"
          >
            <input
              aria-describedby=""
              aria-invalid="false"
              class="css-1w8it3t"
              data-cy="textInput-input"
              id="3"
              type="text"
              value=""
            />
          </div>
          <div
            data-cy="textInput-hintContent"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
